### ISSUE TITLE
Multiple Arguments with object in render breaks cache

### ISSUE THREAD

**Issue by GuiHash** [vercel/swr#150]
Hi,

It seems that multiple arguments with object in render breaks cache : 
```js
// Make sure objects are stable
const params = useMemo(() => ({ id }), [id])
useSWR(['/api/user', params], query)
```

If you follow this use case : 
`id = 1` --> `id = 2` --> `id = 1`
The second `id = 1` doesn't have the same key in cache as the first (because `params` doesn't have the same ref)

Reproductible codesandbox : https://codesandbox.io/s/swr-multiple-arguments-no-cache-mfssd 

The only way I found to preserve cache is to stringify in memo and parse in fetcher (https://codesandbox.io/s/swr-multiple-arguments-cache-p8iwy)
```js
const params = useMemo(() => JSON.stringify({ id }), [id])
```
any better idea ?

Thanks !

**Comment by quietshu** [comment]
Yes this is indeed a problem. In our real world use cases, we're using it like this:
```js
const { data: user } = useSWR('/api/user', fetch)
const { data: userItems } = useSWR(['/api/items', user], fetchWithUser)
```

Because `user` is a globally shared object, the cache key of the second query will always be stable. 
For use cases like query params, I will still suggest passing primitive values as args directly (will update documentation as well):

```js
useSWR(['/api/user', id], (url, id) => query(url, { id }))
```

We have some discussions here https://github.com/zeit/swr/pull/145#discussion_r350064725.

**Comment by sainhimanshu9** [comment]
i have certain params attached with my fetcher, params like pagination limit, offset, filter values and 
 source, the component is reused with different sources the source param changes as the source change, how can i takle this issue?

**Comment by shuding** [comment]
Can you share a code example @sainhimanshu9?

**Comment by ivan-kleshnin** [comment]
@GuiHash `useMemo` is worthless for the described scenario – this hooks does not do real memoization. 
It only compares a new value with an old value so `id = 1`, `id = 2`, `id = 1` sequence will not use "memoized" value.
It will be evaluated 3 times.

**Comment by aej11a** [comment]
@shuding I'm also coming across this problem. Is there any newer recommended solution for this? Our use case is essentially making a generic fetching function which needs to be able to accept flexible variables, so the `passing primitive values as args directly` doesn't fit the use-case.

Maybe useSWR should deep-compare the input, just like it deep compares output? (possibly could be a config option?) Thinking that this would solve the issue you mentioned in your comment on the axios example linked above.

**Comment by ivan-kleshnin** [comment]
@aej11a I use https://www.npmjs.com/package/json-stable-stringify (4 Kb minified, non-Gzipped). Performance-wise it's the same. But I agree that custom compare function would be better given that it's typically already bundled (Lodash/Ramda/etc).

**Comment by sainhimanshu9** [comment]
Hi @aej11a , 
i got the solution, you just need to make a custom hook where you need to pass all params and ( in that custom hook use query-string library) which will return you the query with all  params. hence as the params will change a new query string will be returned.
Mine is working now.

**Comment by aej11a** [comment]
Thanks @ivan-kleshnin and @sainhimanshu9 , for now I'm using a workaround like this:
https://github.com/vercel/commerce/blob/master/framework/commerce/utils/use-data.tsx

In that hook, the variables input is deconstructed and values are used in the query key, and then the fetcher reconstructs the variables object. Seems to be working fine for now, but it's not perfect, so I think first-class library support for deep-comparison for query keys would be better.

**Comment by tadinski** [comment]
> Hi @aej11a ,
> i got the solution, you just need to make a custom hook where you need to pass all params and ( in that custom hook use query-string library) which will return you the query with all params. hence as the params will change a new query string will be returned.
> Mine is working now.

@itsSunnysCode could you share an example of your solution please?

**Comment by nandorojo** [comment]
`react-query` provides deep-equal support for items in an array key. Is this the type of thing we could implement? Maybe a custom `didKeyChange` function could live in the config?

```jsx
<SWRConfig value={{ didKeyChange: (prevKey, nextKey) => deepEqual(prevKey, nextKey) }} />
```

I know that deep-equal checks aren't ideal, and it is slow to run on every render. However, for use-cases with large objects as a key, a plain array key has bad type safety. For instance, I have an algolia search wrapper with swr.

The SWR key takes all of these fields, so I currently use `JSON.stringify`

```ts
const key = [
  'algolia-search',
  query,
  index,
  JSON.stringify({
    facets,
    filters,
    facetFilters,
    numericFilters,
    hitsPerPage,
    page,
    aroundLatLng,
    aroundLatLngViaIP,
    aroundRadius,
    cacheable
  })
]

useSWR(key, algoliaSearch)
```

Passing each options as primitives to an array isn't great. However, stringifying it also isn't very good, since I have to then `JSON.parse` inside of the `fetcher` and cast the types to it:

```ts
const fetcher = async (_, query = '', index, facetString) => {
      const {
        facets = [],
        numericFilters,
        facetFilters,
        filters,
        hitsPerPage,
        page,
        aroundLatLngViaIP,
        aroundLatLng,
        aroundRadius,
        cacheable,
      }: Pick<
        SWRKeyParameters<T>,
        | 'facets'
        | 'filters'
        | 'facetFilters'
        | 'numericFilters'
        | 'hitsPerPage'
        | 'page'
        | 'aroundLatLng'
        | 'aroundLatLngViaIP'
        | 'aroundRadius'
        | 'cacheable'
      > = facetString ? JSON.parse(facetString) ?? {} : undefined
}
```

Is there a type-safe solution that is more efficient than `JSON.stringify`?

**Comment by shuding** [comment]
@nandorojo Yeah we could provide an option to support custom key serialization. There're probably some more efficient JSON stringify implementations, but I'm not sure about type safety...

**Comment by nandorojo** [comment]
Yeah, I also made [this PoC](https://www.typescriptlang.org/play#code/C4TwDgpgBA0hIHkBOBxCxgSQHjiAfFALyzxQQAemAdgCYDOUAFAHRsCGSA5vQFxTtqIAJTFCAS2oAzLFABK6AK5JqANXYAbRdAD8UAHIB7avsUaN7AEYaI2BcGVrN2wvyMmzF67bz4AUH6gkFAAYugAxgAWWLjwADRQACLswOyEJEwA1vD8eMhoGDG+okSEyalQAD5QAApIhgC24vS25WkBQdB4deJNwOIAbtAkgiAA2gC6VVDUntP0wEiSXNNIEOy0xhogAkKTAVKK1OH9xlCKLQDKAOpysTuUNAzMTCWE3Ut9gxCi1R+94n6QwSbXwTD8UCg2RAuXiEKgMmAUSw-DCSOiODwIJS7VEAG8AL4BC4QG5yJgAcki4gpCSYnC4bygePhkL8BOExKutyYrzEUCp8FpzAZTJZkLZHK5pJ5fNKzKgrKg4ikzAAhIttPilRKoMBIvUAO4zCDGgCiSHqSFeSqJurWDhUUDGFJJSHoFKm7EY4WMC3ZCWh-PFEt91AWzvoyIa7CmJGhAV1AHokzNDP1wtB9SkoFHojGKYxOsQiCQ3fQ1ezhEA) to improve type safety for the fetcher function.

If you hover over the first argument in the fetcher function, you'll see that it inherits the types from the `key`.

**Comment by iShawnWang** [comment]
My current working solution : 

```
import useSWR from 'swr';
import { useMemo } from 'react';

export default (key) => {
  const k = useMemo(() => {
    return key;
  }, [JSON.stringify(key)]);
  return useSWR(k);
};

```

**Comment by shuding** [comment]
We now have an example on how to add key-serialization with SWR v1 Middleware: [Serialize Object Keys](https://swr.vercel.app/docs/middleware#serialize-object-keys). This can solve the original issue in this thread, maybe someone can publish it as a handy library?

Deep comparison would be a great addition, but I'm still having concerns about the performance (as well as serialization). Imaging a huge list with `useSWR` hooks inside, each re-render needs to run a synchronized deep comparison. We need to get some benchmarking for this.

**Comment by nandorojo** [comment]
I would love to see some benchmarks for that. Since Algolia doesn’t have React Hooks, I made a hook for it using SWR. The key is really large and complex though.

**Comment by shuding** [comment]
🎉 Added support for key serialization in #1429 and released as `1.1.0-beta.0`, you don't need `useMemo` anymore. CC: @nandorojo
